### PR TITLE
join-pane binding

### PR DESCRIPTION
**Pull request by emanresusername**
seemed more fitting than [here](https://github.com/tmux-plugins/tmux-pain-control/issues/18) as it could cross the session boundary :thinking:

### CODE REVIEW

File: scripts/join_pane.sh [hunk 1]
```diff
@@ -0,0 +1,14 @@
+#!/usr/bin/env bash
+
+CURRENT_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
+
+source "$CURRENT_DIR/helpers.sh"
+
+join_pane() {
+	tmux command-prompt -p "join pane to:" "join-pane -t '%%'"
```
Comment (bruno-) [inline scripts/join_pane.sh]: How do you specify the target pane? Via pane id? Isn't that complex?

File: sessionist.tmux [hunk 2]
```diff
@@ -14,6 +14,9 @@ tmux_option_new="@sessionist-new"
 default_key_bindings_promote_pane="@"
 tmux_option_promote_pane="@sessionist-promote-pane"
 
+default_key_bindings_join_pane="*"
```
Comment (bruno-) [inline sessionist.tmux]: Why not `j` or `J`?

**Comment by bruno-** [comment]
This would be a nice addition to this plugin. However I think the workflow with a prompt is too complex. I propose we use the workflow with marking the current pane (`m` default key binding).

Other things that would need some thinking:

- should `join-pane` join horizontally or vertically? I'd argue that for this plugin that deals with sessions it should NOT split a pane at all. Instead, it should just move the pane to the current session **full screen**.
- Should we still add key bindings for `join-pane` for horizontal and vertical join to this plugin? Should they go to `tmux-pain-control` instead? Maybe this horizontal/vertical join feature isn't that useful and can just be omitted?
- What key bindings to use? My first thought is `j` or `J`, but even though those keys are "free" a lot of people custom-bind them for pane manipulation so that's not the best choice. What about a keybinding with 2 keys? Example `prefix` + `tj`. That might work because `join-pane` feature isn't used as often.

File: scripts/join_pane.sh (hunk 1 above)
Comment (emanresusername) [reply scripts/join_pane.sh]: i tried passing in `#{pane_id}` originally but it never seemed to do the right thing, 
the default (with not specifying `-s` at all) was the version that behaved as expected :confused:

File: sessionist.tmux (hunk 2 above)
Comment (emanresusername) [reply sessionist.tmux]: in a <sup>(hyphenated-)</sup>word: self-interest.  [tmux-pain-control has those](https://github.com/tmux-plugins/tmux-pain-control#bindings) and i use it too :sweat_smile:

**Comment by emanresusername** [comment]
certainly, i assumed this wasn't the cleanest impl, 
was just the best i could do at my current tmux plugin n00b level :sweat_smile: 
could you walk me through that marking work flow ?

for the bottom points

- that makes sense :+1: 
- yeah would make sense to do `pane` joining there and `session` joining here :+1: 
  as for usefulness, i know i'm just 1 data point, but this data point is a resounding **yes** to that! :grin:
- indeed `j` and `J` were taken by `tmux-pain-control` :cry: . 
  with the tmux plugin n00bness, i did not even know 2 keys was an option :astonished:
  works for me :+1: 
  maybe `sj` here (for (s)ession(j)oin) and `pj` and `pJ` in pain control (for (p)ane(j/J)oin, split vertically/horizontally)?

**Comment by bruno-** [comment]
@emanresusername 

> could you walk me through that marking work flow ?

Sure, this is how I currently join panes. It's a manual process, I don't have a keybinding.

- You position the cursor to a pane, hit `prefix + m` (default key bindings) to "mark" that pane.
- Then you navigate to another pane and enter tmux command line mode with `prefix + :` - tmux prompt will open. Then you type `join-pane` and the marked pane will join with the current pane.

> maybe sj here (for (s)ession(j)oin) and pj and pJ in pain control (for (p)ane(j/J)oin, split vertically/horizontally)?

Both keys `s` and `p` are taken by default for important tmux actions.. I don't think it would be smart to mess with those.

**Comment by bruno-** [comment]
@emanresusername If you wanna continue working on this, I think a good first step would be to implement a feature where:

- a user marks a pane with `prefix + m`
- user navigates to another session and hits `prefix + *` (let's leave this as a temporary binding) and the marked pane is joined full screen with the current session.
- user is automatically switched to this joined pane

Similar work was done in "promote pane to session" feature in `scripts/promote_pane.sh`

Let me know how that sounds?

**Comment by emanresusername** [comment]
:+1: i'll take a whack at it this weekend

**Comment by emanresusername** [comment]
some force pushes about to happen you can prolly ignore while i'm trying things out... is there a [way to make `tpm` use a different branch](https://github.com/tmux-plugins/tpm/issues/117) ??

**Comment by emanresusername** [comment]
ok, tinkering pushes done (for now)
this doesn't have as many guards/explicit-pane/session/window-references as the `promote_pane` script
so lmk if it's not safe to assume that what it was defaulting to that was working for me will work for others
BUT
as it stands here, this worked (for me) according to the [3 step flow above](https://github.com/tmux-plugins/tmux-sessionist/pull/21#issuecomment-343137327)

**Comment by bruno-** [comment]
Looking at the solution.. it's super simple. I dunno why I thought this would be more complex.

**Comment by emanresusername** [comment]
yeah, the `mark pane` thing is a game changer! :astonished: 
with that, i see where you were coming from in the comment above:

> Maybe this horizontal/vertical join feature isn't that useful and can just be omitted?

still think it'd be *nice* to make that a *little* easier, but *WAY* less of a hassle already with pane marking

**Comment by bruno-** [comment]
Works very nice, no flicker or anything.. I think the implementation is 👍 

One thing I noticed: if no pane is marked, and the keybinding is pressed it returns a script error. I think we should check this and print a message, like "Mark a pane first".

Implementing horizontal/vertical join should be easy, no? Just pass an argument to the script and branch based on that.

What are your thoughts about the bindings at this point? Here's a couple ideas (I'll omit the  tmux `prefix` for all, that's implied):

- `tj` for join to session, `tJ` for horizontal join, `t<ctrl-j>` for vertical join. The downside of these is I'll have trouble remembering which binding is horizontal and which one is vertical join.
- `t@` for joint to session, `t"` for horizontal join, `t%` for vertical join. I like these because they kinda extend default tmux key bindings (`"` for horizontal pane split) and (`%` for vertical pane split) plus existing feature from this plugin (`@` promote pane into session), but they work with marked pane.

**Comment by emanresusername** [comment]
:+1: to everything in the second bullet and "mark a pane first" message

**Comment by emanresusername** [comment]
[this way](https://web.archive.org/web/20160611073025/http://superuser.com:80/a/424392) was the only way i could find to do the 2-key binding, but i actually kinda like how that ended up working out everything went better than expected

**Comment by emanresusername** [comment]
ok, think that was the last tweak :triumph: 
want me to squash these into 1 commit?

**Comment by bruno-** [comment]
Hey, this is not how you do 2-key bindings. The proper way to do it is explained in `man tmux` under `switch-client`... sorry I assumed this is a known thing..

Here's the example to create a `t*` key binding (I tested and confirmed it's working):

```
tmux bind-key t switch-client -Ttable1
tmux bind-key -Ttable1 '*' join-pane
```

**Comment by emanresusername** [comment]
:clapper: take 3

so this seemed to work for me, and uses the proper `switch-client` method

but i guess the tradeoff is it'll rebind the secondary keys every time in favor of maintaining the style in `sessionist.tmux`

think that's worth it as is? 
or worth breaking the mold in `sessionist.tmux` and doing the secondary key bindings in there as well? 
or something else entirely?

**Comment by bruno-** [comment]
Hey

> but i guess the tradeoff is it'll rebind the secondary keys every time in favor of maintaining the style in sessionist.tmux

Yes, I think all `tmux-plugins` define key bindings when the plugin is loaded in the main plugin file `plugin-name.tmux`. I think that would work here.
My concerns with this approach (defining the bindings every time `switch-client` is done):

- You can't just list keys with `tmux list-keys` and see what bindings are defined.. because current bindings are defined dynamically.
- Dynamically defining key bindings after the first time is redundant.
- Is it slow? I'm on an under-performant computer (from 2013) and some `tmux-plugins` are slow for me.

**Comment by bruno-** [comment]
> 🎬 take 3

You made great progress with this. If you're tired of frustrated with these iterations let me know and I can maybe finish this PR this weekend. That said, I'd rather have you wrap this up because it's your idea and your work.

File: scripts/join_pane.sh [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+# passed to script on the 2-key binding invocation
+JOIN_PANE_FLAG="$1"
+
+source "$CURRENT_DIR/helpers.sh"
+
+join_pane() {
+  local join_pane_flag="$1"
+	tmux join-pane "$join_pane_flag" || \
+    tmux display-message "Mark a(nother) pane first"
```
Comment (bruno-) [inline scripts/join_pane.sh]: I see you're mixing tabs and spaces for indentation on the above 2 lines. Can we use tabs as they're used for all `tmux-plugins`?

**Comment by emanresusername** [comment]
RE: being on a slow computer: i feel ya, mid 2011 laptop checking in :laughing: 
RE: frustrated of the iterations: not at all, all part of the process :+1:

File: scripts/join_pane.sh (hunk 3 above)
Comment (emanresusername) [reply scripts/join_pane.sh]: my editor auto converts things, but i did start playing around with the defaults there part way through

will put in conformant whitespace in the next iteration :+1:

**Comment by emanresusername** [comment]
hopefully that does it :crossed_fingers: 
lmk if you want me to squash these down

**Comment by bruno-** [comment]
Thanks!

**Pull request merged**